### PR TITLE
[SPMD] Introduce schedule_comm_wait

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98580
* #98579
* __->__ #98578
* #98285
* #98284
* #98182

`schedule_comm_wait` delays the wait_tensor ops as late as possible. Note that this optimization currently does not reorder the computation ops. For `foreach` based optimizer, we observe that reordering the computation ops is required to achieve a good performance.

Differential Revision: [D44761487](https://our.internmc.facebook.com/intern/diff/D44761487/)

### CODE REVIEW

File: torch/distributed/_spmd/graph_optimization.py
Comment (fegin) [inline torch/distributed/_spmd/graph_optimization.py:456]: > @mrshenli's internal review: Would I be correct that this optimization will be generalized to other collectives as well?

This pass only looks check the WaitTensor and reorder WaitTensor. So it should be generalized to other collectives  as well.

File: torch/distributed/_spmd/graph_optimization.py
Comment (fegin) [inline torch/distributed/_spmd/graph_optimization.py:468]: > @mrshenli's internal review:  I guess this is assuming none of these users node will be reordered by subsequent graph optimization passes?

IMO, this is a general question for all optimization passes. If some subsequent optimization passes change the topological sort order, there is no guarantee that the previous optimization passes will still have the same effects, regardless the algorithm the previous passes adopt. I added `run_after` argument to the `graph_optimization_pass` contract to expect developers to consider the effect and add the order constraints to the graph optimization passes. But we may need a further discussion on this topic.

File: torch/distributed/_spmd/graph_optimization.py
Comment (fegin) [inline torch/distributed/_spmd/graph_optimization.py:476]: > @mrshenli's internal review:  We definitely need this, but I assume this won't help current dhen use case that much, because those allreduces are schedule right before the first usage in optimizer?

No, the expansion generates allreduce together with WaitTensor. So, without this pass, none of the allreduce will be overlapped with the computation. This was observed in the old _SPMD style expansion for DHEN and the new tracing with a simple MLP model. So I believe we need this pass for DHEN.

**Review by mrshenli — APPROVED** [review]
stamp to unblock. let's test accuracy parity before landing

File: test/distributed/_spmd/test_transformation.py
Comment (fegin) [reply test/distributed/_spmd/test_transformation.py:188]: Done

File: torch/distributed/_spmd/graph_optimization.py
Comment (fegin) [reply torch/distributed/_spmd/graph_optimization.py:472]: `comm_fusion_with_concat` actually moves the allreduce to gradient producers. And this PR move the wait to the gradient consumers. So without reordering the computation op and memory constraint, this is the optimal scheduling.

**Pull request closed without merging**
